File: database_seed/src/entity/KeywordUser.ts
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { Community } from './Community';
import { Keyword } from './Keyword';
import { User } from './User';

@Entity('keyword_user')
export class KeywordUser extends BaseEntity {
  @PrimaryGeneratedColumn('increment', { name: 'id', type: 'bigint' })
  id: string;

  @Column('bigint', { name: 'community_id', nullable: true })
  communityId: string | null;

  @Column('bigint', { name: 'keyword_id', nullable: true })
  keywordId: string | null;

  @Column('bigint', { name: 'user_id', nullable: true })
  userId: string | null;

  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @DeleteDateColumn({ name: 'deleted_at' })
  deletedAt: Date | null;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date | null;

  @ManyToOne(() => Community, (community) => community.keywordUsers, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'community_id', referencedColumnName: 'id' }])
  community: Community;

  @ManyToOne(() => Keyword, (keyword) => keyword.keywordUsers, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'keyword_id', referencedColumnName: 'id' }])
  keyword: Keyword;

  @ManyToOne(() => User, (user) => user.keywordUsers, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'user_id', referencedColumnName: 'id' }])
  user: User;
}

File: rest/src/main/java/com/waglewagle/rest/keyword/entity/KeywordUser.java
package com.waglewagle.rest.keyword.entity;

import com.waglewagle.rest.community.entity.Community;
import com.waglewagle.rest.keyword.data_object.KeywordVO;
import com.waglewagle.rest.user.entity.User;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class KeywordUser {

    @Id
    @GeneratedValue
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "community_id")
    private Community community;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "keyword_id")
    private Keyword keyword;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    private LocalDateTime deletedAt;

    public static KeywordUser of(final KeywordVO.JoinVO joinVO) {
        KeywordUser keywordUser = new KeywordUser();
        keywordUser.community = joinVO.getCommunity();
        keywordUser.keyword = joinVO.getKeyword();
        keywordUser.user = joinVO.getUser();
        return keywordUser;
    }
}
